In [1]:
#pip install nbconvert

In [2]:
#!pip show nbconvert

In [3]:
!pandoc --version

pandoc 3.5
Features: +server +lua
Scripting engine: Lua 5.4
User data directory: C:\Users\ubion\AppData\Roaming\pandoc
Copyright (C) 2006-2024 John MacFarlane. Web: https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.


# 2절 딥러닝으로 AI 모델링하기
인공지능의 급속한 발전을 가능하게 한 것은 컴퓨터 성능의 발전, 딥러닝 알고리즘과 심층 신경망 구조의 혁신, 그리고 빅데이터의 축적입니다. <br>
딥러닝은 인간의 뇌 신경망이 학습하는 방식에서 영감을 얻어 고안되었습니다. 두뇌의 신경세포가 연결된 형태를 모방하여 인공신경망 모델을 만들고 인공신경망을 학습(Learning) 모델로 사용하는 머신러닝 패러다임이 딥러닝(Deep Learning)입니다. <br>
여러 층으로 이루어진 신경망 구조와 다양한 딥러닝 알고리즘은 그동안 많은 연구와 혁신의 결과물입니다. 현재 딥러닝은 이미지 인식, 음성 인식, 언어 번역과 같은 자연어 처리 분야에서 높은 성능을 나타내고 있으며, 언어 번역, 고객 상담, 자율주행, 질병 진단 등 딥러닝의 활용 분야가 지속적으로 확장되고 있습니다. <br>
이번 섹션에서는 기본적인 인공신경망 개념과 여러 층으로 이루어진 심층신경망을 효과적으로 학습하는 알고리즘을 소개하고, 분류 문제를 위한 간단한 심층신경망을 알아봅니다. 

## 1.인공신경망
두뇌의 신경망 정보처리 구조를 모방하여 만든 컴퓨터 게산 알고리즘이 인공신경망(Arificial Neural Network, ANN)입니다. 두뇌의 신경망은 수많은 뉴런(neuron, 신경세포)이 연결되어 정보를 처리하고 전달합니다. <br>
생물학적 뉴런에는 외부 자극(신호)을 수용하는 여러 개의 수상돌기가 있으며, 세포핵에서 이 자극들을 새로운 자극으로 가공합니다. 새로운 자극은 축삭돌기를 통해 전달됩니다. <br>
생물학적 뉴런을 인공 뉴런으로 모델링하면, 입력 $x_1$, $x_2$, $x_3$ ⋯ $x_n$에 가중치 $w_1$, $w_2$, $w_3$ ⋯ $w_n$를 곱하고, 이를 모두 더해서 활성화 함수를 통과시키면 출력됩니다. 







일반적으로 인공 뉴런을 노드(node) 또는 유닛(unit)이라고 하며, 인공신경망은 노드들의 그룹으로 연결되어 뇌의 신경망 구조와 유사합니다. 인공신경망에서 원 모양은 노드를 나타내고 화살표는 하나의 노드 출력에서 다른 노드로 입력을 나타냅니다. 인공신경망의 입력과 출력 중간에 있는 층을 은닉층(hidden layer)이라고 합니다. 

#### 1) 가중치와 편향 이해하기
인공신경망 모델을 수학적으로 표현하면 다음 그림과 같습니다. 뉴런에 2개의 입력값 , 이 입력되고, 입력 개수만큼 2개의 가중치(weight) $w_1$, $w_2$와 1개의 편향(bias, 바이어스) $b$를 가지고 있습니다. 

가중치는 입력 데이터에서 원하는 연산 결과를 만들도록 비중을 부여하는 값이고, 편향은 뉴런의 활성화를 조절하는 상수입니다. 가중치와 편향은 뉴런의 동작 특성을 나타내는 중요한 파라미터로서 이 값들을 조정해 원하는 출력값을 만들어 냅니다. <br>
인공신경망으로 실제 정답에 근사하는 출력값을 만들기 위해 뉴런의 가중치와 편향을 반복적으로 조정하며, 이러한 반복적인 과정을 학습이라고 합니다. <p>
인공신경망의 가중합을 구하는 수식은 간단한 1차 함수로 표기할 수 있으며, 입력 에 가중치 를 각각 곱하고, 그 값을 모두 더한 후에 추가적으로 편향을 더합니다. 

$$가중합=\sum{}_{i}w_1x_1+b=(w_1x_1+w_2x_2)+b$$
$$(W : 가중치, b : 편향)$$


입력 $x_1$이 매우 큰 값이라면 $x_1$이 연산 결과에 미치는 영향도가 절대적일 것입니다. 하지만 입력 $x_1$에 곱하는 가중치 $w_1$이 0.0001처럼 작은 값이라면, $w_1$×$x_1$의 연산 결과는 0에 가까운 값이 됩니다. 즉, 가중치는 입력 데이터의 연산 결과에 미치는 영향도를 조절하는 요소이고, 편향값 $b$도 가중합을 조정하는 파라미터입니다. 

### 2) 활성화 함수 알아보기
생물학적 뉴런은 입력된 자극(신호)이 특정 강도 이상일 때만 다음 뉴런으로 신호를 전달하는데, 인공신경망의 뉴런에서도 동일한 역할을 하는 활성화 함수(activation function)가 있습니다.<br>
활성화 함수는 입력값들의 수학적 선형 결합인 가중합 값을 입력받고 일정 기준에 따라 변환하여 값을 출력하는 비선형(또는 선형) 함수입니다. <br>
주요 활성화 함수는 시그모이드(sigmoid), 하이퍼볼릭 탄젠트(hyperbolic tangent, tanh), 렐루(ReLU)가 있으며, 최근 인공신경망의 은닉층에서는 렐루를 많이 사용합니다. 

#### (1) 시그모이드
시그모이드 함수의 수식은 다음과 같습니다. 
$$
\sigma(x) = \frac{1}{1 + e^{-x}} \qquad f'(x) = f(x) \cdot (1 - f(x))
$$

활성화 함수 $f(x)$와 활성화 함수의 미분 결과 $f'(x)$를 그래프로 나타내면 다음과 같습니다. <br>
시그모이드 함수는 입력값을 비선형 형태로 0~1 사이의 값으로 변화시키므로, 로지스틱 회귀와 값은 이진 분류 모델의 결과를 확률적으로 나타내는 데 사용합니다. 


시그모이드 미분 함수 $f'(x)$의 작은 입력값이 0일 때 0.25에 불과하고 입력 $x$가 매우 작아지거나 커지면 기울기(미분값)가 0에 수렴하는 것을 알 수 있습니다. 인공신경망을 사용하는 딥러닝은 모델학습에 활성화 함수의 기울기(gradient, 그래디언트)를 계속 곱하는 연산이 있으며, 미분값이 작은 시그모이드 함수를 은닉층의 활성화 함수로 사용하면 기울기 소실(vanishing gradient) 문제가 발생해 모델 학습이 제대로 이루어지지 않습니다. 

#### (2) 하이퍼볼릭 탄젠트
시그모이드 활성화 함수를 사용하는 경우 은닉층이 많아질수록 효과적으로 학습되지 않는 한계점을 개선하기 위한 방법 중 하나로 하이퍼볼릭 탄젠트 함수가 제안되었습니다. 
하이퍼볼릭 탄젠트 함수의 출력은 –1과 1 사이의 값이고, 미분값이 시그모이드 함수에 비해 커졌습니다. 하지만 하이퍼볼릭 탄젠트 함수도 $x$값이 크거나 작아짐에 따라 기울기가 매우 작아져서 기울기 소실 문제가 발생합니다.

$$
f(x) = \tanh(x) = \frac{sinhx}{coshx} = \frac{e^x - e^{-x}}{e^x + e^{-x}}
$$
$$
e : 지수함수(exponential function)$$



#### (3) 렐루
렐루 함수는 시그모이드와 하이퍼볼릭 탄젠트 활성화 함수가 갖는 기울기 소실 문제를 해결하기 위한 함수로 현재 가장 많이 사용되는 활성화 함수 중 하나입니다. 
렐루 함수는 입력값이 양수일 경우 출력값은 입력값 $x$와 같고, 입력값이 음수일 경우 출력값은 0으로 렐루 함수의 수식은 다음과 같습니다. 

$$f(x) = \max(0, x) \quad f(x) = \begin{cases} 0 & \text{if } x < 0 \\ x & \text{if } x \geq x \end{cases} \quad f'(x) = \begin{cases} 0 & \text{if } x < 0 \\ 1 & \text{if } x \geq x \end{cases}$$<p>


그래프를 보면 입력값이 양수일 경우 미분값은 항상 1이며, 기울기 소실 문제를 해결할 수 있습니다. 그리고 렐루 함수에는 특별한 연산이 없으므로 연산 속도가 빠릅니다. 


렐루 함수에도 단점이 있는데, 렐루 함수의 입력값이 음수일 경우 미분값이 항상 0이므로, 입력값이 음수인 뉴런의 출력을 회생시키지 못하는 한계가 있습니다. 




#### (4) 리키 렐루
렐루 함수의 단점을 보완하기 위해 다양하게 변형된 렐루 함수가 제안되었습니다. 그중 리키 렐루는 입력값이 음수일 때 출력값을 0이 아닌 0.001과 같은 매우 작은 값을 출력하는 함수로 수식은 다음과 같습니다.
$$f(x) = max(ax,x)$$
$a$는 0.01, 0.001과 같이 작은 값으로 설정하는 파라미터입니다. 리키 렐루 함수는 입력값 0 이하에서도 기울기가 0이 되지 않아 뉴런의 출력값이 사라지는 현상을 방지할 수 있습니다. 

#### (5) 소프트맥스
소프트맥스 함수는 다중 분류 모델에 사용하는 활성화 함수로 수식은 다음과 같습니다. 

$$
\text{Softmax}(\mathbf{z}) = \frac{e^{z_i}}{\sum^k_{j=1} e^{z_i}}\qquad \text for \,i = 1, 2,\,…\,k
$$


소프트맥스 함수는 모든 입력값을 0과 1 사이의 값으로 정규화하여 출력하며, 출력값들의 총합은 1이 됩니다. 분류할 클래스가 n개라고 할 때, 소프트맥스 출력값은 n개이며, 이는 입력이 각 클래스에 속할 확률 추정치입니다. 
예를 들어 분류할 클래스가 3개인 경우 소프트맥스 함수의 결과는 다음과 같습니다. 


수식은 조금 복잡하지만, 풀어서 설명하면 소프트맥스 함수는 분류하고자 하는 클래스가 3개일 때, n차원의 벡터를 입력받아서 합은 1이 되도록 모든 벡터 원소의 값을 0과 1 사이의 값으로 변경하여 3차원의 벡터를 반환하는 함수입니다. 
다음 예시와 같이 값들이 소프트맥스 함수를 통해 0.71, 0.05, 0.24와 같은 0과 1 사이의 값으로 변환되고, 변환된 값들의 총합은 1입니다. 소프트맥스 함수의 출력은 확률값이 되고, 이 중 가장 큰 확률값이 분류의 결과가 되며, 예시에서는 분류 결과가 클래스 A로 되었습니다. 

## 2. 심층신경망
### 1) 심층신경망 구조 알아보기
입력층(input layer)과 출력층(outpyt layer) 사이에 여러 개의 은닉층(hidden layer)이 있는 인공신경망을 심층신경망(Deep Neural Network, DNN)이라고 합니다. 

심층신경망은 입력 데이터를 수신하는 입력층과 여러 개의 은닉층, 그리고 모델의 출력이 나오는 출력층으로 구성됩니다. <br>
심층신경망에서 은닉층과 노드(뉴런)의 개수가 많아질수록, 입력 데이터 자체만으로는 미처 알지못한 높은 수준의 특징과 패턴을 찾아낼 수 있습니다. 심층신경망의 은닉층에서는 입력 데이터에서 높은 수준의 특성/패턴 데이터를 추출하고, 심층신경망의 출력층에서 분류/예측 모델이 해결하고자 하는 문제에 대한 알고리즘이 적용됩니다. <br>
딥러닝 알고리즘을 이해하기 위해서는 다음 용어들을 알아야 합니다. 
<div style="display:table; border-collapse:collapse; width:100%; text-align:center;">
    <div style="display:table-row; background-color:#d9e2f3; font-weight:bold">
        <div style="display:table-cell; border:1px solid black; padding:5px;">용어</div>
        <div style="display:table-cell; border:1px solid black; padding:5px;">설명</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">샘플</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">예측에 사용하는 데이터로, 입력(input)이라고도 표기합니다.</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">입력</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">모델이 규칙을 찾아야 할 대상으로, 샘플(sample)과 같은 용어 입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">타깃</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">모델이 예측해야 하는 대상으로, 샘플(입력)에 상응하는 레이블(label)입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">학습</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">컴퓨터가 스스로 데이터의 규칙을 찾아내는 기술입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">훈련</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">데이터의 규칙을 찾고 정확도를 높이는 방향으로 수정하는 과정, 즉 학습을 구현하기 위한 과정입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">알고리즘</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">특정 문제를 해결하거나 계산을 수행하는 데 필요한 단계의 순서를 구체적으로 명시하는 규칙의 조합으로, 딥러닝에서 알고리즘은 입력 데이터를 학습해 ‘예측’이라는 결과를 출력하는 것을 의미합니다.</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">모델</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">데이터 학습이 완료된 알고리즘으로, 수학식이나 샘플 데이터를 입력받아서 타깃 레이블 값을 예측하는 함수입니다. 딥러닝을 통해 최종적으로 얻고자 하는 산출물입니다. </div>
    </div>    
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">파라미터</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">모델을 규정하는 값으로 가중치와 편향이 있습니다. </div>        
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">예측</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">모델치를 추정(estimate)하는 값입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">손실</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">모델의 출력값과 타깃인 실제값 간의 차이를 비교하는 함수입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">손실함수</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">모델의 예측인 출력값과 타깃인 실제값 간의 차이를 비교하는 함수입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">경사하강법</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">어떤 손실함수가 정의되었을 때, 손실함수의 값이 최소가 되는 지점을 찾아가는 방법입니다. </div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px;">오차역전파</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">신경망에서 순전파 이후에 출력값에 오차가 있는 경우, 출력층에서 입력방향으로(역방향) 오차가 작아지도록 가중치가 더 이상 없데이트 되지 않을 때까지 반복하는 알고리즘입니다. </div>
    </div>
</div>

### 2) 신경망 출력 계산하기
심층신경망의 출력 계산은 입력층에서 시작해서 정방향으로 훈련 데이터의 패턴을 신경망의 네트워크에 전파하는 순전파(forward propagation) 과정으로 이루어집니다. <br>
신경망 출력 계산 방법을 설명하기 위해 $w^1_{12}$과 같은 표기를 사용합니다. 표기의 위첨자는 신경망의 몇 번째 층인지를 나타내는 인덱스 번호입니다. 그리고 표기의 아래첨자 앞자리 숫자는 앞 층에서 몇번 째 뉴런인지를, 아래첨자 뒷자리 숫자는 다음 층에서 몇 번째 뉴런인지를 나타내는 인덱스 번호입니다. <br>
데이터의 패턴이 정방향으로 신경망의 네트워크로 전파될 때, 현 단계 뉴런의 가중치와 전 단계 뉴런의 출력값 곱들의 합을 입력값으로 받습니다. 이 같은 다시 활성화 함수를 통해 다음 뉴런으로 전파됩니다. <br>
최종적으로 출력층에서 나오는 값은 이 모델이 예측한 결괏값입니다. 심층신경망을 출력을 기반으로 손실함수를 이용하여 오차를 계산하고, 심층신경망 네트워크에 있는 모든 가중치에 대한 도함수를 찾아서 오차를 역전파하여 모든 가중치와 편향을 업데이트합니다. <br>
반복적인 순전파와 역전파를 통해 최적의 편향값들을 찾았다면 모델 학습이 잘된 것이며, 해당 출력을 모델의 예측값으로 사용할 수 있습니다. 

### 3) 손실함수 이해하기
손실함수는 신경망 학습의 목적 함수로 신경망의 출력값(예측값)과 실제값의 차이를 계산하는 함수입니다. 회귀 모델과 분류 모델에 주로 사용하는 손실함수를 알아봅니다. 

#### (1) 회귀 모델 손실함수
수치를 예측하는 회귀 모델(Regression)은 손실함수로, 평균제곱오차(Mean Squared Error, MSE)를 주로 사용합니다. 평균제곱오차(MSE)는 수학적인 분석이 쉽고 계산이 용이합니다.<br> 
평균제곱오차는 실제값과 예측값의 차이인 오차들의 제곱 평균이어서 특이치에 민감합니다. 예측값과 실제값의 절댓값 차이인 평균절대오차(MAE)도 회귀 모델의 손실함수로 사용하며, MAE는 MSE보다 이상치에 덜 민감합니다.

#### (2) 분류 모델 손실함수
카테고리를 분류하는 분류 모델(Classification)의 손실함수는 크로스 엔트로피(Cross Entropy)함수를 사용합니다. 크로스 엔트로피 함수는 실제값(레이블)과 모델의 예측값이 많이 틀릴수록, 그에 비례하여 더 큰 손실값이 부여되도록 디자인된 함수입니다. 




분류 모델의 종류에는 이진 분류와 다중 분류가 있습니다. 이진 분류는 True 또는 False, 양성 또는 음성 등 2개의 클래스로 분류하는 모델이며, 이진 교차 엔트로피 오차(Binary Cross Entropy Error, BCEE) 함수를 사용합니다. 이진 교차 엔트로피 오차함수 공식은 다음과 같습니다. 

$$
L = - \frac{1}{N} \sum_{i=1}^{N} \left[ y_i \log(\hat{y}_i) + (1 - y_i) \log(1 - \hat{y}_i) \right]
$$

여러 개의 레입르 중 하나로 분류하는 다중 분류 모델 훈련에는 범주형 교차 엔트로피 오차 함수를 사용합니다. 다중 분류는 분류해야 할 클래스가 3개 이상인 경우로 범주형 교차 엔트로피 오차함수의 공식은 다음과 같습니다. 

$$
L = - \frac{1}{N} \sum_{i=1}^{N} \sum_{c=1}^{C} y_{i,c} \log(\hat{y}_{i,c})
$$



#### 4) 경사하강법 이해하기
딥러닝에서 학습 데이터를 입력하여 신경망 구조를 거쳐 결괏값(예측값)을 얻습니다. 모델이 예측한 값과 실제값의 차이인 오차를 최소화하는 파라미터를 찾는 과정이 최적화이며, 신경망 최적화기법으로 경사하강법(Gradient Descent)이 대표적입니다. <br>
경사하강법은 오차가 최소화되도록 파라미터를 업데이트하는 데 손실함수의 기울기(gradient)를 사용합니다. <br>
다음 그래프와 같이 손실함수는 가중치(w)에 대한 함수로, 손실함수가 볼록함수 형태라면 미분으로 손실이 가장 작은 가중치를 찾을 수 있습니다. 하지만 딥러닝에서는 손실함수가 복잡하고 계산량이 매우 크며 미분이 0이 되는 값이 여러 개 존재하므로 미분만으로 최솟값을 찾기 어려워 경사하강법을 사용합니다. 아래 그림에서 보듯이, 경사하강법은 손실함수의 현재 가중치에서 기울기를 구해 손실을 줄이는 방향으로 가중치를 업데이트해 나갑니다. 


경사하강법의 단점을 보완한 방법으로 확률적 경사하강법, 미니배치 경사하강법이 있습니다. 경사하강법은 정확하게 가중치를 찾아가지만 파라미터를 한번 업데이트하는 데 데이터세트 전체를 사용합니다. 딥러닝 모델의 학습에서 사용하는 데이터세트 규모는 보통 수백만 개가 넘는데, 많은 계산량 때문에 속도가 느리거나 최적해를 찾기 전에 학습을 멈출 수 있는 단점이 있습니다. <br>
확률적 경사하강법은 파라미터를 업데이트하기 위해 무작위로 샘플링된 1개의 데이터를 사용하는 방법이며, 적은 계산량으로 적절한 기울기를 구할 수 있지만 노이즈가 심하다는 단점이 있습니다. <br>
미니배치 경사하강법은 랜덤 추출한 일부 데이터를 통해 가중치를 조절합니다. 예를 들어 학습 데이터가 1,000개이고 배치 크기를 100으로 하면, 총 10개의 미니배치가 나오고, 이 미니배치 하나 당 한 번씩 경사하강법을 진행하여 속도를 개선합니다. 

경사하강법의 최적화 기법에는 관성의 효과를 내는 모멘텀(Momentum), 파라미터의 업데이트 횟수에 따라 학습률을 조절하는 옵션이 추가된 아다그라드, 지수 이동평균을 이용하는 알엠에스프롭, 모멘텀과 알엠에스프롭을 융합한 방법인 아담 등 다양합니다. 특히 아담은 모멘텀과 알엠에스프롭을 섞어놓은 최적화 알고리즘으로 딥러닝에서 자주 사용됩니다. 

### 5) 역전파 알고리즘 이해하기
신경망 학습 방법은 모델의 출력값과 실제값의 차이인 오차값을 최소화하기 위해 모델의 파라미터를 조정하는 오차역전파 알고리즘을 사용합니다. <br>
심층신경망의 출력값은 각 노드들에서 계산되는 가중합들의 연산 결과이므로 가중치와 편향이 모델에서 변경 가능한 파라미터이고, 오차값이 적은 출력값을 만드는 최적의 가중치들과 편향값들을 찾는 것이 딥러닝 모델의 목적입니다. <br>
처음에는 심층신경망 각 노드(뉴럽)의 가중치와 편향을 무작위로 부여하고, 순전파와 오차역전파를 계속 반복하는 방식으로 가중치들과 편향값들을 조정하면서 오차를 최소화합니다. 

다음 그림과 같이 어떤 입력에 대한 실제값이 10인 경우, 신경망의 정방향 계산에 의한 출력값이 7이라면 오차는 실제값 10에서 신경망의 출력값 7을 뺀 3이 됩니다. 

출력층에서 측정한 오차를 이전의 은닉층과 입력층으로 역방향 전파하면서 가중치들과 편향값을 조정합니다. 이때 역방향으로 전파되는 오차의 크기는 각 뉴런이 결괏값에 미치는 영향도에 비례합니다. 예를 들어 가중치 $w^2_{11}$에는 1.5, $w^2_{12}$에는 1.2, $w^2_{13}$에는 0.3만큼의 오차를 줄일 수 있도록 가중치 값을 조정합니다. 

**■ 역전파 알고리즘 동작 방식**<br>
① 신경망 모델은 모든 가중치(w)와 편향(b)을 랜덤하게 초기화합니다.<br>
② 신경망 모델에 x값을 입력하면 신경망은 예측 y를 출력합니다. <br>
③ 신경망에서 손실 최소화를 위해 가중치를 업데이트해야 합니다. <br>
④ 역전파를 사용해 신경망에 있는 모든 가중치에 대한 비용함수의 기울기를 계산합니다. <br>
⑤ 가중치에 대한 비용함수의 기울기에 비례하여 가중치를 조정함으로써, 역전파는 비용을 감소하는 방향으로 가중치를 변경할 수 있습니다. 

#### 6) 드롭아웃 이해하기
보통 사용하는 심층신경망은 파라미터가 매우 많은 모델입니다. 파라미터가 많으면 모델의 복잡도가 높은 것이고, 모델의 복잡도가 너무 높으면 과적합(overfitting) 경향이 발생합니다. 심층신경망이 훈련 데이터를 외워버리면 훈련 데이터세트에서는 좋은 성능을 내지만, 별도의 테스트 데이터세트나 새로운 데이터세트에서는 성능이 낮습니다.<br> 
드롭아웃(Dropout)은 심층신경망에 적용하여 과대적합을 방지하고 일반화 성능을 높이는 방법입니다. 드롭아웃은 심층신경망을 훈련하는 동안, 반복마다 은닉 유닛의 일부를 확률 p만큼 랜덤하게 드롭해서 학습에 참여하지 않도록 하는 방법입니다. 드롭아웃 확률은 사용자가 지정하고, p=0.5를 많이 사용하여 이 경우 50%의 은닉 노드가 랜덤하게 드롭됩니다. 데이터를 신경망에 순전파하는 과정에서 은닉 노드가 생략되므로 역전파에서도 제외됩니다. 일반화는 모델 훈련에 사용된 데이터가 아닌 이전에 접하지 못한 새로운 데이터에 대한 올바른 에측을 수행하는 능력을 의미합니다. 

## 3. 딥러닝 프레임워크
### 1) 텐서 이해하기
텐서(Tensor)는 다차원 배열을 통칭하는 용어입니다. 스칼라는 값을 1개의 수치로 표현하고, 벡터는 하나의 값을 2개 이상의 수치로 표협합니다. 그리고 매트릭스는 2개 이상의 벡터 값을 통합해 구성된 값입니다. <br>
행렬을 2차원의 배열이라고 한다면 텐서는 2차원 이상의 다차원을 가진 배열로 이해하면 됩니다. 

#### 2) 딥러닝 개발 프레임워크 알아보기
딥러닝 모델을 만들고 사용하기 위한 대표적인 프레임워크에는 텐서플로, 케라스, 파이토치가 있으며, 오픈소스로 제공됩니다. <br>
딥러닝 프레임워크는 딥러닝에 필요한 많은 알고리즘을 제공하기 때문에, 이를 활용할 경우 복잡한 로직이나 수학 연산을 직접 구현할 필요 없이 솔루션 개발에 집중할 수 있습니다. 

**■ 텐서플로(TensorFlow)**<br>
- 구글이 개발하여 오픈소스로 공개한 딥러닝 프레임워크로 데이터 플로 그래프(Data Flow Graph) 구조를 사용하여 수치 연산을 합니다. <br>
- 강력한 시각화 기능과 높은 수준의 모델 개발에 사용할 수 있는 여러 옵션을 갖춘 딥러닝 라이브러리입니다. <br>
- 프로덕션 레디 배포(production-ready deployment) 옵션과 모바일 플랫폼 지원에 장점이 있습니다. <br>
- 텐서플로 홈페이지：https://www.tensorlfow.org

**■ 케라스(Keras)**<br>
- 텐서플로의 상위 레발 API로 딥러닝 모델을 쉽게 구현하는데 도움을 주는 직관적인 API를 제공합니다. <br>
- 시퀀셜(Sequential) 모델로 원하는 레이어(layer)를 쉽게 순차적으로 쌓을 수 있습니다. <br>
- 다중 출력 등 복잡한 모델을 케라스 함수 API로 쉽게 구성할 수 있습니다. <br>
- 케라스 홈페이지：https://keras.io

**■ 파이토치(PyTorch)**<br>
- 페이스북이 개발한 파이썬 기반 오픈소스 라이브러리 토치(Torch)를 기반으로 만들어진 프레임워크입니다.<br>
- 동적 계산 그래프를 사용하며 파이썬 친화적인 프레임워크입니다.<br>
- 파이토치 홈페이지：https://pytorch.org
  
#### 3) 텐서플로 설치 및 활용하기
딥러닝 모델 개발 실습에 텐서플로를 사용해 봅니다. 파이썬이 설치된 환경에서 파이썬 라이브러리 관리 툴 pip로 텐서플로를 설치합니다. <br>
- 터미널/명령창：pip install tensorflow<br>
- 주피터 노트북：!pip install tensorflow

주피터 노트북에서 텐서플로를 설치합니다. 

In [4]:
#!pip install tensorflow

설치가 완료되면 다음 코드로 텐서플로 버전을 확인합니다.

In [5]:
# 설치된 텐서플로 버전 확인하기
import tensorflow as tf
print(tf.__version__)

2.13.0


먼저 필요한 라이브러리를 불러오고(import) 딥러닝 모델 학습에 사용할 데이터를 생성합니다. 딥러닝 모델에 훈련 데이터로 사용하는 데이터 x_train은 넘파이 reshape() 메소드로 데이터의 형태를 바꿔줍니다. 

In [6]:
# 필요한 라이브러리 불러오기
import numpy as np
import tensorflow as tf
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

# 모델 학습 데이터 생성하기
x = [1, 2, 3, 4,  5, 6,  7,  8,  9, 10]
y = [3, 5, 7, 9, 11, 13, 15, 17, 19, 21]
x_train = np.array(x)
x_train = x_train.reshape(-1, 1)
y_train = np.array(y)

print(f'입력 데이터 : {x_train}')
print(f'입력 데이터 형태: {x_train.shape}')
print(f'출력 데이터: {y_train}')
print(f'출력 데이터 형태 : {y_train.shape}')

입력 데이터 : [[ 1]
 [ 2]
 [ 3]
 [ 4]
 [ 5]
 [ 6]
 [ 7]
 [ 8]
 [ 9]
 [10]]
입력 데이터 형태: (10, 1)
출력 데이터: [ 3  5  7  9 11 13 15 17 19 21]
출력 데이터 형태 : (10,)


케라스의 Sequential 클래스로 층(layer)을 선형으로 연결하는 신경망을 구성합니다.<p>
- Sequential 객체를 생성하고 add() 메소드로 층을 추가합니다.<br>
- Sequential 모델의 첫 번째 층은 input_shape 또는 input_dim 인자로 입력 형태에 대한 정보를 받습니다.  두 번째 이후 충돌은 앞선 층의 출력과 연결되기 때문에 자동으로 형태를 추정할 수 있으므로 형태 정보를 갖고 있을 필요는 없습니다. <br>
- 케라스 layers API에서는 Dense layer, Activation layer, Conv2D layer 등 다양한 층을 제공합니다. 여기에서는 기본적인 Dense 층을 사용하며, Dense 층은 한 층의 유닛이 다른 층의 모든 유닛과 연결된 상태를 의미합니다. Dense 클래스의 주요 매개변수들은 다음과 같습니다. <br>
  - units：출력 유닛 개수를 설정합니다. <br>
    - input_shape：입력의 개수를 설정합니다. <br>
  - activation：활성화 함수를 설정합니다. <br>
  - liner：기본값으로 입력값과 가중치로 계산된 결괏값이 그대로 출력됩니다. <br>
  - sigmoid：스기모이드 함수로 이진 분류 출력층에 사용합니다. <br>
  - softmax：소프트맥스 함수로 다중클래스 분류 문제 출력층에 사용합니다. <br>
  - relu：렐루(Rectified Linear Unit) 함수로 은닉층에서 주로 사용합니다. <p>

Conv 2D 등 다른 레이어에 대한 정보는 케라스 API 페이지(https://keras.io/api/layers)를 참고합니다. 

그리고 summary() 메소드로 신경망의 구성을 확인할 수 있습니다. 

In [7]:
# Keras의 Sequential 모델 구성하기
initializer = tf.keras.initializers.GlorotUniform(seed=42) #모델 시드 고정하기
model = Sequential()
model.add(Dense(units=1, input_shape=(1,),kernel_initializer=initializer))
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 1)                 2         
                                                                 
Total params: 2 (8.00 Byte)
Trainable params: 2 (8.00 Byte)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


compile 메소드로 모델의 학습에 대한 환경을 설정합니다. <br>
compile 메소드의 매개변수는 다음과 같습니다. <br>
- optimizer：optimizer 알고리즘으로 sgd, adam 등이 있습니다.<br> 
- loss：손실함수로 mse, categorical_crossentropy 등이 있습니다. <br>
- metric：모델의 성능을 평가하는 데 사용되는 함수로 mae, acc 등이 있습니다. 

In [8]:
# 모델을 학습시킬 최적화 방법, loss 계산 방법, 평가 방법 설정하기
model.compile(optimizer='sgd', loss='mse', metrics=['mae'])

모델을 학습시키기 위해서는 일반적으로 fit 함수를 사용합니다. 

fit 메소드의 인자는 다음과 같습니다. <br>
- x：입력데이터<br>
- y：타깃 데이터(label)<br>
- batch_size：한 번에 학습할 때 사용하는 데이터 개수로 batch_size가 10이라면 10개의 데이터를 학습한 다음 가중치를 한번 갱신합니다. <br>
- epochs：학습 데이터 반복 횟수로 학습 데이터세트 전체를 몇 번 학습하는지를 의미합니다. 학습 데이터세트를 여러 번 학습할수록 학습 효과는 커집니다. 하지만 학습을 많이 하면 모델의 가중치가 학습 데이터에 지나치게 최적화되는 과대적합 현상이 발생합니다. 

In [9]:
# 모델 학습하기
model.fit(x_train, y_train, epochs=1000)

Epoch 1/1000
1/1 [==============================] - 0s 444ms/step - loss: 323.2251 - mae: 16.1452
Epoch 2/1000
1/1 [==============================] - 0s 9ms/step - loss: 14.9011 - mae: 3.5555
Epoch 3/1000
1/1 [==============================] - 0s 7ms/step - loss: 0.7542 - mae: 0.8584
Epoch 4/1000
1/1 [==============================] - 0s 9ms/step - loss: 0.1045 - mae: 0.2802
Epoch 5/1000
1/1 [==============================] - 0s 10ms/step - loss: 0.0741 - mae: 0.2255
Epoch 6/1000
1/1 [==============================] - 0s 8ms/step - loss: 0.0721 - mae: 0.2238
Epoch 7/1000
1/1 [==============================] - 0s 8ms/step - loss: 0.0715 - mae: 0.2230
Epoch 8/1000
1/1 [==============================] - 0s 9ms/step - loss: 0.0709 - mae: 0.2223
Epoch 9/1000
1/1 [==============================] - 0s 6ms/step - loss: 0.0703 - mae: 0.2214
Epoch 10/1000
1/1 [==============================] - 0s 10ms/step - loss: 0.0697 - mae: 0.2205
Epoch 11/1000
1/1 [==============================] - 0s 11ms/

학습이 완료된 모델의 가중치를 확인합니다. 

In [10]:
# 모델 가중치 확인하기
print(model.weights)

[<tf.Variable 'dense/kernel:0' shape=(1, 1) dtype=float32, numpy=array([[2.001265]], dtype=float32)>, <tf.Variable 'dense/bias:0' shape=(1,) dtype=float32, numpy=array([0.99119264], dtype=float32)>]


모델 내 각 레이어의 가중치와 편향을 개별적으로 학인할 수 있습니다.

In [11]:
# 모델 레이어의 가중치 출력하기
print(f'weight : {model.layers[0].weights[0].numpy()}')
print(f'bias : {model.layers[0].bias.numpy()}')

weight : [[2.001265]]
bias : [0.99119264]


학습이 완료된 모델을 사용하고 예측하고 결과를 확인해 봅니다. 

In [12]:
# 학습 완료된 모델 사용하여 예측하기
print(model.predict([[11],[12],[13]]))

1/1 [==============================] - 0s 95ms/step
[[23.005108]
 [25.006372]
 [27.007637]]


학습이 완료된 모델의 가중치와 편향을 대입하여 선형함수로 표현하면 다음과 같습니다. 

$$y=2.001265*x + 0.99119264$$
모델의 출력은 입력 데이터에 2를 곱하고 1을 더한 값과 비슷하게 나온다는 것을 알 수 있습니다. help 함수로 클래스(객체)와 메소드의 사용법을 출력할 수 있으므로, 자세한 사용법을 help 함수로 확인합니다. 

다음 예제는 hlep를 사용하여 확인한 fit() 메소드 내용입니다. 

In [13]:
# 클래스와 메소드 사용법 확인하기
help(model.fit)

Help on method fit in module keras.src.engine.training:

fit(x=None, y=None, batch_size=None, epochs=1, verbose='auto', callbacks=None, validation_split=0.0, validation_data=None, shuffle=True, class_weight=None, sample_weight=None, initial_epoch=0, steps_per_epoch=None, validation_steps=None, validation_batch_size=None, validation_freq=1, max_queue_size=10, workers=1, use_multiprocessing=False) method of keras.src.engine.sequential.Sequential instance
    Trains the model for a fixed number of epochs (dataset iterations).
    
    Args:
        x: Input data. It could be:
          - A Numpy array (or array-like), or a list of arrays
            (in case the model has multiple inputs).
          - A TensorFlow tensor, or a list of tensors
            (in case the model has multiple inputs).
          - A dict mapping input names to the corresponding array/tensors,
            if the model has named inputs.
          - A `tf.data` dataset. Should return a tuple
            of either `(

## 4. 심층신경망으로 항공사 고객 만족 분류 모델 구현 실습하기
심층신경망으로 항공사 이용 고객의 만족 여부를 분류하는 모델을 구현해 봅니다. 

### 1) 데이터 불러오기 및 확인하기
데이터는 항공사 고객 만족 예측 데이터(Airline Customer Satisfaction)의 ‘Inveistico_Airline.csv’를 사용합니다.<br>
- 데이터 출처：https://www.kaggle.com/datasets/sjleshrac/airlines-customer-satisfaction<br>
- 데이터명：invisitico_Airline.csv

**■ 데이터 설명**<br>
항공사 고객 12만 9,880명의 설문 결과와 인구통계/개인정보로 구성되어 있으며, 데이터에는 총 23개의 속성이 있습니다. 먼저 9개의 변수를 확인해 봅니다. 

<div style="display:table; border-collapse:collapse; width:80%; text-align:center;">
    <div style="display:table-row; background-color:#d9e2f3; font-weight:bold">
        <div style="display:table-cell; border:1px solid black; padding:5px;">칼럼명</div>
        <div style="display:table-cell; border:1px solid black; padding:5px;">설명</div>
        <div style="display:table-cell; border:1px solid black; padding:5px;">비고</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Satisfaction</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">고객 만족 여부(Y)</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Satisfied : 만족<br>Dissatisfied : 불만족</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Gender</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">성별</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Female : 여성<br>Male : 남성</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Customer Type</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">고객 유형</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Loyal Customer : 충성 고객<br>Disloyal Customer : 비충성 고객</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Age</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">나이</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">리뷰를 받은 고객의 나이</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Type of Travel</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">여행 유형</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Business travel : 비즈니스 여행<br>Personal Travel : 개인 여행</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Class</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">좌석 등급</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Business : 비즈니스<br>Eco : 이코노믹<br>Eco Plus: 이코노믹 플러스</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Flight Distance</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">비행거리</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">&nbsp;</div>
    </div>    
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Departure Delay in Minutes</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">출발 지연 시간(분)</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">&nbsp;</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Arrival Delay in Minutes</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">도착 지연 시간(분)</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">&nbsp;</div>
    </div>
</div>
다음 14개의 변수는 고객의 만족도 항목으로 0에서 5 사이의 척도 점수를 매깁니다. 
<div style="display:table; border-collapse:collapse; width:40%; text-align:center;">
    <div style="display:table-row; background-color:#d9e2f3; font-weight:bold">
        <div style="display:table-cell; border:1px solid black; padding:5px;">칼럼명</div>
        <div style="display:table-cell; border:1px solid black; padding:5px;">설명</div>
    </div>
    <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Seat comfort</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">좌석의 편안함</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Departure/Arrival time convenient</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">편리한 출발/도착 시간</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Food and drink</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">음식과 음료</div>
    </div>
     <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Gate location</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">게이트 위치</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Inflight entertainment</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">기내 와이파이 서비스</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Online support</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">기내 엔터테인먼트</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Online support</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">온라인 지원</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Ease of Online booking</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">온라인 예약의 용이성</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">On-board service</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">온보드 서비스</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Leg room service</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">레그룸 서비스</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Baggage handling</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">수화물 처리</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Checkin service</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">체크인 서비스</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Cleanliness</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">청결</div>
    </div>
        <div style="display:table-row;">
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">Online boarding</div>
        <div style="display:table-cell; border:1px solid black; padding:5px; text-align:left;">온라인 탑승</div>
    </div>
</div>

필요한 라이브러리를 불러옵니다. 

In [14]:
import numpy as np
import pandas as pd
import warnings

warnings.filterwarnings('ignore')

Pandas library의 read_csv() 메소드를 사용해, csv 파일에서 데이터를 불러오고 데이터 프레임(DataFrame)으로 저장합니다.<br>
infor() 메소드로 데이터프레임의 정보를 확인합니다. 

In [15]:
# csv 파일에서 데이터를 로드해서 데이터프레임으로 저장하기
df = pd.read_csv('Invistico_Airline.csv')

# 데이터프레임 정보 확인하기
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129880 entries, 0 to 129879
Data columns (total 23 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   satisfaction                       129880 non-null  object 
 1   Gender                             129880 non-null  object 
 2   Customer Type                      129880 non-null  object 
 3   Age                                129880 non-null  int64  
 4   Type of Travel                     129880 non-null  object 
 5   Class                              129880 non-null  object 
 6   Flight Distance                    129880 non-null  int64  
 7   Seat comfort                       129880 non-null  int64  
 8   Departure/Arrival time convenient  129880 non-null  int64  
 9   Food and drink                     129880 non-null  int64  
 10  Gate location                      129880 non-null  int64  
 11  Inflight wifi service              1298

#262

데이터를 살펴봅니다. 

In [16]:
# 데이터프레임의 처음 5개 행의 데이터 출력하기
df.head()

satisfaction  Gender   Customer Type  Age   Type of Travel     Class  \
0    satisfied  Female  Loyal Customer   65  Personal Travel       Eco   
1    satisfied    Male  Loyal Customer   47  Personal Travel  Business   
2    satisfied  Female  Loyal Customer   15  Personal Travel       Eco   
3    satisfied  Female  Loyal Customer   60  Personal Travel       Eco   
4    satisfied  Female  Loyal Customer   70  Personal Travel       Eco   

   Flight Distance  Seat comfort  Departure/Arrival time convenient  \
0              265             0                                  0   
1             2464             0                                  0   
2             2138             0                                  0   
3              623             0                                  0   
4              354             0                                  0   

   Food and drink  ...  Online support  Ease of Online booking  \
0               0  ...               2                       3   
1               0  ...               2                       3   
2               0  ...               2                       2   
3               0  ...               3                       1   
4               0  ...               4                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 3                 0                 3                5   
1                 4                 4                 4                2   
2                 3                 3                 4                4   
3                 1                 0                 1                4   
4                 2                 0                 2                4   

   Cleanliness  Online boarding  Departure Delay in Minutes  \
0            3                2                           0   
1            3                2                         310   
2            4                2                           0   
3            1                3                           0   
4            2                5                           0   

   Arrival Delay in Minutes  
0                       0.0  
1                     305.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 23 columns]

In [17]:
데이터의 통계량을 확인합니다. 

SyntaxError: invalid syntax (1872631259.py, line 1)

In [ ]:
df.describe()

각 칼럼별로 결측치가 있는지 확인합니다. 

In [ ]:
df.isnull().sum()

### 2) 데이터 전처리하기
데이터를 확인한 결과, 도착 지연 시간(​Arrival Delay in Minutes) 칼럼에 Null 값이 393개 있으며, 텍스트로 구분되는 범주형 칼럼이 다수 있습니다. 신경망 모델의 입력 데이터에는 결측치가 없어야 하고, 기본적으로 수치형 데이털르 사용합니다. 그러므로 범주형 데이터를 수치형 데이터로 변환하는 인코딩 작업을 하며, 수치형 데이터도 모델 성능을 높이기 위해 데이터 스케일링 등의 변환 작업이 요구됩니다. 
#### (1) 결측치 처리하기
사이킷런 SimpleImputer 객체를 사용해서 도착 지연 시간(​Arrival Delay in Minutes) 칼럼에 있는 결측치를 평균값으로 치환합니다. 

In [ ]:
#!pip install scikit-learn

In [ ]:
# SimpleImputer 객체로 결측치 대체하기
from sklearn.impute import SimpleImputer

mean_imputer = SimpleImputer(strategy='mean')
df["Arrival Delay in Minutes"] = mean_imputer.fit_transform(df[["Arrival Delay in Minutes"]])

#### (2) 데이터 인코딩하기
Object 칼럼의 유형을 String 유형으로 변경합니다. 

In [ ]:
# object 칼럼 유형을 string 유형으로 변경하기
cols = ['satisfaction', 'Gender', 'Customer Type', 'Type of Travel', 'Class']
df[cols] = df[cols].astype(str)

범주형 데이터를 수치값으로 변경합니다. 

In [ ]:
# 범주형 데이터를 수치값으로 변경하기
df['satisfaction'].replace(['dissatisfied','satisfied'], [0,1], inplace=True)

좌석 등급(Class) 칼럼은 순서를 고려해 정수 1~N으로 순서형 인코딩(Ordinal Encoding)을 합니다. ‘Eco’, ‘Eco Plus’, ‘Business’ 값이 0, 1, 2로 변환됩니다. 순서형 인코딩의 경우 판다스의 Categorial 함수를 활용합니다. 

In [ ]:
# 순서형 인코딩(Ordinal Encoding)하기
categories = pd.Categorical(
    df['Class'], 
    categories= ['Eco', 'Eco Plus', 'Business'], 
    ordered=True)
labels, unique = pd.factorize(categories, sort=True)
df['Class'] = labels

범주형 데이터인 성별(Gender), 고객 유형(Customer Type), 여행 유형(Type of Travel) 칼럼은 원핫 인코딩을 적용합니다. 원핫 인코딩은 0과 1의 벡터로만 표현하는 기법입니다. 범주(Gategory)의 수만큼 벡터의 수가 생성되므로, 각 범주가 새로운 특성이 됩니다. 

In [ ]:
# 원핫 인코딩(One Hot Encoding)하기
cat_cols = ['Gender','Customer Type','Type of Travel']
df = pd.get_dummies(df, columns=cat_cols)

데이터 전처리 결과를 확인합니다. 모든 데이터 유형이 수치형 데이터로 변경되었습니다.

In [ ]:
df.head()

In [ ]:
df.dtypes

#### (3) 데이터세트 분리하기
데이터세트를 입력(X)과 레이블(y)로 분리하고, 훈련 데이터세트(train dataset)와 검증 데이터세트(validation dataset)로 분리합니다. 데이터세트는 test_size에 지정한 비율로 분리됩니다. test_size를 0.2로 지정했으므로, 훈련 데이터는 80%, 검증 데이터는 20%의 비율로 데이터세트가 분리됩니다. 

In [ ]:
from sklearn.model_selection import train_test_split

# 데이터셋을 입력(X)과 레이블(y)로 분리하기
X = df.drop(['satisfaction'], axis=1)
y = df['satisfaction'].reset_index(drop=True)

# 데이터셋을 훈련 데이터와 검증 데이터로 분리하기
X_train, X_val, y_train, y_val = train_test_split(X, y, 
    test_size=0.2, 
    random_state=42, 
    stratify=y)

print(f'훈련 데이터셋 크기 : X_train {X_train.shape}, y_train {y_train.shape}')
print(f'검증 데이터셋 크기 : X_val {X_val.shape}, y_val {y_val.shape}')

#### (4) 데이터 스케일링하기
특성별로 데이터의 스케일이 다르면 딥러닝이 잘 동작하지 않을 수 있습니다. 따라서 데이터 스케일링 작업을 통해 모든 특성의 범위(또는 분포)를 유사하게 만들어줘야 합니다. 사이킷런 MinMaxScaler 객체로 데이터의 최솟값과 최댓값을 이용하여 데이터를 특정 범위(주로 0~1 사이)로 스케일링하여 특성을 정규화합니다. 사이킷런의 StandardScaler 객체도 스케일링에 많이 사용합니다. StandardSaler는 특성들의 평균을 0, 분산을 1로 스케일링하여 정규분포로 만드는 표준화를 합니다. <br>
이번 실습에서는 MinMaxScaler 객체를 생성하여 fit 메소드로 학습시킨 후 transform 메소드를 사용하여 데이터를 변환합니다. 

In [ ]:
from sklearn.preprocessing import MinMaxScaler

# 데이터 정규화하기
scaler = MinMaxScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_val = scaler.transform(X_val)

print(X_train)

### 3) 심층신경망 모델 생성하기
입력 데이터는 25개, 은닉층은 여러 개, 출력은 1개인 이진 분류를 위한 심층신경망(DNN, Deep Neural Network) 모델을 구성합니다. 은닉층의 활성화 함수는 ‘relu’를 사용하고, 마지막 층의 활성화 함수는 출력이 1개인 이진 분류 모델이므로 ‘sigmoid’를 사용합니다. 

In [ ]:
# 필요한 라이브러리 불러오기
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.layers import Input, Dense, Dropout
from tensorflow.keras.layers import BatchNormalization 
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
import random

# 모델 시드 고정하기
tf.random.set_seed(42)
np.random.seed(42)
random.seed(42)

# Keras의 Sequential 객체로 딥러닝 모델 구성하기
initializer = tf.keras.initializers.GlorotUniform(seed=42) #모델 시드 고정하기
model = Sequential()
model.add(Dense(32, activation='relu', input_shape=(25,),kernel_initializer=initializer))
model.add(Dense(64, activation='relu'))
model.add(Dense(128, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dropout(0.3))
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(1, activation='sigmoid'))

모델 구조와 파라미터의 개수를 확인해 보면, 모델은 총 8개의 층으로 구성되었고 학습해야 할 총 파라미터 개수(Total params)는 2만 1,633개입니다. 

In [ ]:
# 모델 구조 및 파라미터 정보 확인하기
model.summary()

#### 4) 모델 컴파일하기
optimizer는 adam, 모델 성능 평가 메트릭으로는 정확도를, 이진분류 모델이므로 손실함수는 binary_crossentropy를 사용합니다. 

In [ ]:
# 모델을 학습시킬 최적화 방법, loss 계산 방법, 평가 방법 설정하기
model.compile(optimizer='adam', 
              loss='binary_crossentropy', 
              metrics=['accuracy']) 

#### 5) 모델 학습하기
모델 학습 시 너무 많은 에포크(epoch)는 과대적합을 발생시킵니다. 이를 방지하기 위한 방법으로 조기 종료(Early Stopping)가 있습니다. 조기 종료는 검증 데이터세트에서 성능이 더 이상 증가하지 않으면 모델 학습을 중단하는 방법입니다. EarlyStopping 객체를 생성하여 model.fit 함수의 callback 매개변수에 넣어주면 조기 종료가 적용됩니다. EarlyStopping의 주요 인자는 다음과 같습니다. 
- monitor：학습 조기 종료를 위한 성능 모니터링 도구로, val_loss나 val_accuracy가 주로 사용되며 기본값은 val_loss입니다. 
- min_delta：개선되고 있다고 판단하기 위한 최소 변화량입니다. 
- patience：성능 향상을 몇 번의 에포크 동안 기다릴지 설정하며 기본값은 0입니다.
- verbose：얼마나 자세하게 정보를 표출할 것인가를 지정하며, 가능한 값은 0, 1, 2입니다. 
- mode：성능 모니터링 도구의 개선 판단기준으로, monitor  설정 항목이 val_loss이면 min을 설정하고, val_accuracy이면 max를 설정해야 합니다. 기본값은 auto로 monitor 설정된 항목에 따른 판단 기준을 자동으로 지정합니다. 
- restore_best_weights：관할 항목의 가장 좋은 값을 가지는 에포크의 모델 가중치 복원(restore) 여부입니다. 기본값은 False로, 학습 에포크의 마지막 가중치를 보존합니다. 

예제 코드에서 validation loss가 10번 이상 개선되지 않으면 학습을 중단하고, 가장 성능이 좋았을 때의 가중치를 사용합니다. epochs는 100으로 설정했고, 훈련 과정의 loss, accuracy를 history에 저장합니다. 

In [ ]:
# 모델 학습하기
es = EarlyStopping(monitor='val_loss', min_delta=0, patience=10, verbose=1, restore_best_weights=True)
history = model.fit(X_train, y_train, epochs=100, batch_size=128,
          verbose=1, validation_data=(X_val, y_val), callbacks=[es])

### 6) 모델 훈련 과정 시각화하기
신경망 모델의 훈련에 사용되는 fit()메소드는 history 객체를 반환합니다. history.history 속성은 모델의 훈련 과정에서 에포크(epoch)에 따른 정확도(accuracy)와 같은 성능 지표와 손실값을 기록합니다. 그리고 검증 지표와 손실값도 기록합니다.<br> 
예시 코드는 딥러닝 모델의 훈련 과정에서 기록된 정확도(accuracy), 검증 정확도(val_accuracy), 손실(loss), 검증 손실(val_loss)을 그래프로 출력하여 시각화합니다. 

In [ ]:
!pip install tensorflow

In [ ]:
import matplotlib.pyplot as plt

# 훈련 과정 정확도(accuracy) 시각화하기
plt.plot(history.history['accuracy'])
plt.plot(history.history['val_accuracy'])
plt.title('Model Accuracy')
plt.xlabel('Epoch')
plt.ylabel('Accuracy')
plt.legend(['Train', 'Validation'], loc='lower right')
plt.show()

# 훈련 과정 손실(loss) 시각화하기
plt.plot(history.history['loss'])
plt.plot(history.history['val_loss'])
plt.title('Model Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend(['Train', 'Validation'], loc='upper right')
plt.show()

## 2. 기계 학습(ML, Machine Learning)
### 1) 정의: 
기계 학습은 데이터로부터 패턴을 학습하여 새로운 데이터에 대해 예측하거나 의사결정을 내리는 알고리즘을 의미합니다.<br>
사람이 일일이 프로그램을 작성하지 않아도, 컴퓨터가 데이터에서 스스로 학습하여 성능을 개선하는 것이 목표입니다. 

**학습(Training)이란,**<br>
- 데이터를 기반으로 모델이 규칙을 찾아내고, 이를 통해 새로운 데이터를 예측할 수 있는 능력을 갖추는 과정을 의미합니다. 
- 이 과정에서 모델은 주어진 데이터를 바탕으로 패턴을 학습하고, 이를 통해 미래의 데이터를 예측하거나 분류하는 능력을 갖추게 됩니다. 
### 2) 사례: 
이메일 스팸 필터링, 사용자 맞춤형 추천 시스템(Netflix나 YouTube의 콘텐츠 추천), 질병 진단 예측 모델 등이 기계 학습의 활용 예입니다. 
### 3) 특징:
기계 학습은 지도 학습(Supervised Learning), 비지도 학습(Unsupervised Learning), 강화 학습(Reinforcement Learning) 등의 다양한 학습 방법을 포함합니다. 데이터를 이용해 모델을 학습시키고, 학습된 모델은 새로운 데이터를 처리하고 예측하는 데 사용됩니다. 기계 학습은 주로 통계적 모델과 알고리즘을 기반으로 패턴을 인식하고, 이를 바탕으로 의사 결정을 내립니다.

![이미지 설명](./001-04.png)

